### PR TITLE
Sorting

### PR DESCRIPTION
**Pull request by sidnarayanan**
Two changes to sorting:
- `Collection::sort` only makes a temporary copy of `array_` now, not the data itself. There may be a way to improve this further by using move semantics, but I haven't quite gotten it working yet.
- `RefVector::sort` is now implemented. 

As per the suggestion in #101 , I've also added sorting to the release validation mechanism. Actually, it's a bit more general than that, but the only thing implemented so far is pT sorting. It looks like:
```muons/MuonCollection(32)  // relvalsort=pt otheraction=otherargument```

This is stored in `BranchObject.modifier`, if present. These modifiers can be used elsewhere for customization. `relvalsort` triggers the addition of an initialization string to the `plotter` class of the first branch of the collection (since we only want to test the sorting once).

### CODE REVIEW

File: Framework/interface/Iterator.h [hunk 1]
```diff
@@ -17,28 +17,42 @@ namespace panda {
...
-      bool operator==(const_type const& rhs) const { return ptr_.obj == rhs.ptr_.obj; }
-      bool operator!=(mutable_type const& rhs) const { return ptr_.obj != rhs.ptr_.obj; }
-      bool operator!=(const_type const& rhs) const { return ptr_.obj != rhs.ptr_.obj; }
+
       reference operator*() const { return *ptr_.obj; }
       pointer operator->() const { return ptr_.obj; }
+
       int operator-(self_type const& rhs) const { return ptr_.obj - rhs.operator->(); }
-      self_type operator[](int n) const { return this->operator+(n); }
+      self_type& operator[](int n) { auto itr(*this); return (itr += n); }
```
Comment (yiiyama) [inline Framework/interface/Iterator.h]: I was still unsure about this but I think I know what it should be now. In this [example from GNU libstdc++](https://gcc.gnu.org/onlinedocs/libstdc++/libstdc++-html-USERS-4.3/a02021.html), the return type of `__normal_iterator::operator[]` is `reference`. Indeed if you think of a normal pointer as a random access iterator,
```
char* s = "hello world";
char& c = s[4]; // <- c is 'o'
```
works. And the idea is to replicate this behavior with `iterator::operator[]`.

So long story short, this line should be
```
reference operator[](int n) { auto itr(*this); return *(itr += n); }
```

File: Framework/interface/RefVector.h [hunk 2]
```diff
@@ -151,6 +163,21 @@ namespace panda {
...
+  {
+    // define an index comparison using the element comparison
+    auto indexComp([this, &_c](short i1, short i2)->Bool_t {
+        return _c(static_cast<Element const&>((*container_)->elemAt(i1)),
+                  static_cast<Element const&>((*container_)->elemAt(i2)));
+      });
+
+    std::sort(indices_->begin(), indices_->end(), indexComp);
+
+    return *indices_;
```
Comment (yiiyama) [inline Framework/interface/RefVector.h]: If we are just returning `*indices_`, it doesn't provide any additional information, so I think we should rather just return nothing (`indices_` is always accessible from the `indices()` function). By dereferencing and returning an rvalue, there is an implicit copy of index vector taking place.

**Comment by yiiyama** [comment]
Hey Sid, I added a few comments - sorry to be a pain in the butt, but since you've been spending quite some time on making this proper and beautiful, I think we should carry it all the way through.

Also, nice idea about the modifier and sorting as a specific case. Just FYI though, Most of the PandaProd XYZFiller does sort objects at the end by pT in decreasing order already.

File: Framework/interface/Iterator.h (hunk 1 above)
Comment (sidnarayanan) [reply Framework/interface/Iterator.h]: In that case, it's safe to make this function `const`, right? It's no longer possible to do assignment through this subscript (although you can modify the value directly, but I don't think that violates the `const`-ness of the iterator).

File: Framework/interface/RefVector.h (hunk 2 above)
Comment (sidnarayanan) [reply Framework/interface/RefVector.h]: Ah sure. I had left the return value in to mirror `Collection::sort`, but I can drop it.

File: Framework/interface/Iterator.h (hunk 1 above)
Comment (sidnarayanan) [reply Framework/interface/Iterator.h]: It looks like this is what `__normal_iterator` does:
```C++
       reference
       operator[](const difference_type& __n) const
       { return _M_current[__n]; }
```

Comment (yiiyama) [reply Framework/interface/Iterator.h]: Yes it can be (and is in the example) const indeed.

File: RelVal/interface/EnumerateBranches.h [hunk 3]
```diff
@@ -1663,6 +1663,7 @@ template <>
 struct plotter <124> {
   constexpr static const char* name = "muons/size";
   std::vector<float> operator () (panda::Event& event) {
+    event.muons.sort(panda::Particle::PtGreater);
```
Comment (dabercro) [inline RelVal/interface/EnumerateBranches.h]: I was thinking more along the lines that we could assert that the relevant output vector is already correctly sorted. So we'd have something like `assert(is_sorted(output))` in the struct with name `"muons/pt"`.

Comment (dabercro) [reply RelVal/interface/EnumerateBranches.h]: Though we might want to be a little less aggressive than a literal `assert`, and change the exit code accordingly.

**Comment by sidnarayanan** [comment]
Just realized this was never merged...do we need further discussion/testing?

**Comment by dabercro** [comment]
My usual approach is to merge if no one comments for a week.
